prov/efa: Add initial unit test implementation

wckzhang opened the pull request:
Signed-off-by: William Zhang <wilzhang@amazon.com>

shijin-aws commented on prov/efa/test/README.md:
````diff
@@ -0,0 +1,61 @@
+To run efa unit tests, you will need to have cmocka installed.
+See: https://cmocka.org/
+
+You will need to configure libfabric with ```--enable-efa-unit-test```.
````
A high level question: Does this unit test require efa device? If I configure libfabric with such flag and `make check` on a node that does not have efa device, will the test fail?

a-szegel replied:
Tests must have changed... now they do.

a-szegel replied:
Currently, an EFA device is required.

wckzhang commented:
The appveyor failure looks legitimate though I'm not sure why it can't find the header file

wckzhang commented:
Added header check for cmocka.h and cleaned up the autotools code a bit

wckzhang commented:
Looks like I screwed something up while cleaning up the autotools, let me fix that

wckzhang commented:
Should be fixed now

wzamazon commented on prov/efa/src/efa_static_headers.h:
```diff
@@ -0,0 +1,19 @@
+ *
+ */
+
+int rxr_info_to_rxr(uint32_t version, const struct fi_info *core_info,
```
Instead of testing `rxr_info_to_rxr`, I think `rxr_getinfo`  is a better target, which is the entry point of application, and  the input and output is well defined. (or even `rxr_prov->get_info`, if testing a struct is possible)

wzamazon commented on prov/efa/src/rxr/rxr_ep.c:
```diff
@@ -1896,7 +1896,7 @@ static inline void rxr_ep_check_peer_backoff_timer(struct rxr_ep *ep)
  * @param[in]	cqe_to_process	max number of cq entry to poll and process
  */
-static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep,
+unit_static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep,
```
if we are not testing `rdm_ep_poll_ibv_cq` then we do not need this change.

wckzhang replied:
I can remove this change

wzamazon commented on prov/efa/test/efa_test_getinfo.c:
```diff
@@ -0,0 +1,97 @@
+                           const struct fi_info *util_hints,
+                           struct fi_info **core_info);
+
+int __wrap_rxr_get_lower_rdm_info(uint32_t version, const char *node,
```
Instead of mocking `rxr_get_lower_rdm_info`. We need to mock functions like `ibv_open_device`.  `ibv_query_device`, `ibv_query_gid`. Because the behavior is well defined.

wckzhang replied:
Sure

wzamazon commented on prov/efa/test/efa_test_getinfo.c:
```diff
@@ -0,0 +1,97 @@
+	ret = rxr_info_to_rxr(0, core_info, info, hints);
+	assert_int_equal(ret, 0);
+	assert_int_equal(info->caps, rxr_info.caps);
+	assert_int_equal(info->mode, rxr_info.mode);
```
The first 2 tests does not do much, and can be ignored. I would really like to see a test case that we use mock function to return multiple EFA devices.

a-szegel replied:
@wzamazon Is this still blocking the PR from being merged?

wckzhang commented on Makefile.am:
```diff
@@ -15,6 +15,7 @@ AM_CPPFLAGS = \
 
 noinst_LTLIBRARIES =
 lib_LTLIBRARIES =
+check_PROGRAMS =
```
can't add to check programs if not initialized

wckzhang commented:
Removed static mocking until further needed. Changed testing to use getinfo tests.

Wanted to do some mocking with ibv_get_device_list, but this runs on symbol loading...which makes it impossible to manipulate as a per unit test function. We'll have to fix values as far as I can tell.

wckzhang commented:
@shefty can you let me know what the intel failure is?

shefty commented:
No idea what the Intel failure was, just said aborted.  Maybe the CI nodes were getting sleepy.  Restarted it.

a-szegel commented on prov/efa/test/efa_unit_tests.c:
```diff
@@ -0,0 +1,14 @@
+	cmocka_set_message_output(CM_OUTPUT_XML);
+
+	ret = cmocka_run_group_tests_name("efa getinfo tests", getinfo_tests, NULL, NULL);
+	if (ret != 0)
```
Is there any reason for this if statement?

wckzhang replied:
At some point I wrote an example that was supposed to be "tests that fail" and had this check, but I can remove it.

a-szegel commented on prov/efa/Makefile.include:
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
+if ENABLE_EFA_UNIT_TEST
+check_PROGRAMS += prov/efa/test/efa_unit_test
+TESTS += prov/efa/test/efa_unit_test
+nodist_prov_efa_test_efa_unit_test_SOURCES = \
```
I could see us wanting multiple UT executables, if our UT's are doing things that corrupt "normal" state of the system (i.e.. impact each other), but I would rather have 1 main executable that holds all of our test with an excluded list for tests that need their own executable (vs the current implementation's include list).

wckzhang replied:
I'm not sure if we can do that. We might want to run the same tests under different configurations, such as "one nic" and "four nics".

It is possible to do a wildcard and add all files in a directory, but that's generally suggested against with autotools and I'm not sure that wildcarding is entirely system agnostic. In general, file additions should be uncommon enough that manual steps, or at most a script that adds the files to a Makefile.am is all that is needed.

wckzhang commented:
> Can I mock any internal function in libfabric with the current setup (common or provider code), i.e.. things that are in the executable, or are we relying on a linker trick, and only supporting mocking functions in external shared libraries?

This should be able to mock any internal function as long as you include the appropriate headers.

> We should keep all of our mocks in one common place ex: prov/efa/test/scaffolding. We should strive to develop a common mock library, which can be utilized by all unit tests. Instead of pre-developing the mock library, I think that we should make a template mock + directions, to be copied/pasted by dev's who are creating new mocks for their UT's. The template will encourage developers to build their new mocks in a common way which will drive reuse for our tests, and populate our mock library as we go.

This is doable

wzamazon commented on prov/efa/configure.m4:
```diff
@@ -113,6 +115,35 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
+				AC_DEFINE([EFA_UNIT_TEST], [1],
+					[EFA unit testing])
+				efatest_CPPFLAGS="$efatest_CPPFLAGS $efadv_CPPFLAGS"
+				efatest_LIBS="-lcmocka -libverbs -static $efatest_LIBS"
```
Do not you need find cmocka and setup include path?

wzamazon commented on prov/efa/Makefile.include:
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
+
+prov_efa_test_efa_unit_test_LDADD = $(efatest_LIBS) $(linkback)
+prov_efa_test_efa_unit_test_CPPFLAGS = -I$(top_srcdir)/include -pthread $(efatest_CPPFLAGS) -I$(top_srcdir)/prov/efa/src -I$(top_srcdir)/prov/efa/src/rxr -I$(top_srcdir)/prov/efa/test
+prov_efa_test_efa_unit_test_LDFLAGS = $(efatest_LDFLAGS) -Wl,--wrap=rxr_info_to_rxr,--wrap=ibv_get_device_list
```
`rxr_info_to_rxr` is not needed any more. right?

wckzhang replied:
Yeah let me remove that

a-szegel commented on prov/efa/Makefile.include:
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
 	prov/efa/src/rxr/rdm_proto_v4.h
 
+
+if ENABLE_EFA_UNIT_TEST
```
When I configure with for unit testing, I still need to manually add the cmocka library path to my LD_LIBRARY_PATH for `make check` to work.  The output of `make check` is dense and hard to parse visually, and I think the tester needs to `cat prov/efa/test/efa_unit_test.log` to see what went wrong.

a-szegel replied:
Fixing this with --cmocka <path>, and adding that to rpath.

mmubarak commented on prov/efa/test/README.md:
```diff
@@ -0,0 +1,62 @@
+You can use the cmocka API to change the behavior of your mock function.
+
+The will_return class of functions will place values onto a stack that can
+be popped within the function with the mock function. You can use this to
```
@wckzhang added a few comments regarding README in this PR: https://github.com/ofiwg/libfabric/pull/7472

a-szegel commented on prov/efa/src/efa.h:
```diff
@@ -111,6 +114,12 @@
+#ifdef EFA_UNIT_TEST
+#define unit_static
+#else
+#define unit_static static
```
OBE, Team decided to put the UT's in the same file as the src, so we can unit test static functions now.

a-szegel commented on prov/efa/Makefile.include:
```diff
@@ -73,9 +73,26 @@ _efa_headers = \
 
+
+if ENABLE_EFA_UNIT_TEST
+check_PROGRAMS += prov/efa/test/efa_unit_test
```
Talked with Wei, he would rather run `make check` on a compute node in the cluster, or manually attach a EFA Device to the head of the cluster.

wckzhang commented:
Continuing this PR in PR #7472
